### PR TITLE
🌌 Resolve tracking config templates

### PR DESCRIPTION
**Pull request by shantanuraj**
## :star2: What does this PR do?

- 🧠 Add custom deserialization for BranchConfig
- 🕴️ Store branch config as array
- 🌌 Add custom parsing for TrackConfig
- 🏋️‍♂️ Add `toggl config active`

## :bug: Recommendations for testing

- See config parsing demo.
  [asciicast](https://asciinema.org/a/qQXexPPkAKHXULne9JM1t8on4)
- Another demo showing the config matching via the `toggl config active` command
  [asciicast](https://asciinema.org/a/r6fBXDNbA0DLEcTMwfaTW8Oai)
- Try running `toggl config` for various macros and shell commands.
- Sample complex configuration with a bunch of shell expansions

```toml
['*']
description = "🔨 {{branch}} / {{current_dir}}"
project = "{{parent_base_dir}}"
billable = true

['alt']
description = "chore(release)"
project = "{{base_dir}}"
tags = ["release", "{{parent_dir}}"]
billable = false

# Match all branches that start with a number and a slash
# e.g. 1234/feature-branch
['^\d+/.*']
description = "🔨 {{branch}}"
project = "{{parent_dir}}/{{base_dir}}"
# Set task to android if the current directory contains "android"
# otherwise set it to iOS
task = """
  {{$if [[ "$PWD" == *"android"* ]]; then echo "Android"; else echo "iOS"; fi}}
"""
# Set tags to the current year
tags = ["code", "{{$ date +\"%Y\" }}"]
billable = true
```

### New commands introduced

```console
# Get active tracking configuration for current directory and branch
$ toggl config active
```

## :memo: Links to relevant issues/docs

See #45 which defines the marcro expansion mechanism.

## :speech_balloon: Summarise how you feel about this PR with a gif/image

negroni-sbagliato

### CODE REVIEW

**Review by heytherewill — CHANGES_REQUESTED** [review]
This is amazing!

One thing I'd add: Add something in the lines of `[Currently active]` when printing the list of branch configs so that the user only needs the `toggl config active` command if they have a massive config file

File: src/config/model.rs [hunk 1]
```diff
@@ -63,3 +244,275 @@ impl std::fmt::Display for TrackConfig {
...
+                Ok(output) => {
+                    if !output.status.success() {
+                        println!(
+                            "{}: {}\n{}: {}",
+                            "Error resolving shell macro".red(),
+                            String::from_utf8(output.stderr)?.red().bold(),
+                            "Command".yellow(),
+                            command.yellow().bold()
+                        );
+                        return Ok(command);
```
[the remark this answers is not in the document]
Comment (shantanuraj) [reply src/config/model.rs]: So I've cleaned all surrounding code and added proper errors for config related errors in the error crate.

https://github.com/watercooler-labs/toggl-cli/commit/e6441dc1ed52dd18ed9839314302226d9104c692
https://github.com/watercooler-labs/toggl-cli/pull/53/commits/0aa99257257311bf3fd7b5ed9cd42f3d701e01ce

And in https://github.com/watercooler-labs/toggl-cli/commit/1f4dd231c89886b0d318cbe03710b5a1a2c1b8a9 I updated the parser to ignore config values whose macros cannot be substituted.

File: src/config/model.rs [hunk 2]
```diff
@@ -63,3 +244,275 @@ impl std::fmt::Display for TrackConfig {
...
+                        .to_str()
+                        .unwrap()
+                        .trim()
+                        .to_string())
+                }
+                Err(err) => Err(Box::new(err)),
+            }
+        }
+        Macro::Shell(command) => {
+            let output = std::process::Command::new("sh")
```
[the remark this answers is not in the document]
Comment (shantanuraj) [reply src/config/model.rs:431]: I'll need to check with you to run some commands on windows to resolve this one.

File: src/config/model.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shantanuraj) [reply src/config/model.rs]: Examples for

## Invalid macro
image

## Shell resolution error
image

**Pull request merged**